File: integrations/langgraph/python/ag_ui_langgraph/a2ui_tool.py
"""
A2UI subagent tool factory for LangGraph agents.

Thin adapter over ``ag-ui-a2ui-toolkit`` — the heavy lifting (op builders,
prompt assembly, history walkers, output envelope) lives in the toolkit so
each new framework adapter (ADK, Mastra, Strands, …) only owns the
framework-specific glue: tool decorator, runtime state access, model
binding + invoke.

Example usage in a chat node::

    from ag_ui_langgraph import get_a2ui_tools

    a2ui = get_a2ui_tools(model=ChatOpenAI(model="gpt-4o"))

    model_with_tools = chat_model.bind_tools(
        [*state["tools"], a2ui],
        parallel_tool_calls=False,
    )
"""

from __future__ import annotations

from typing import Any, Optional

from langchain.tools import tool, ToolRuntime
from langchain_core.language_models import BaseChatModel
from langchain_core.messages import SystemMessage

from ag_ui_a2ui_toolkit import (
    A2UI_OPERATIONS_KEY,
    BASIC_CATALOG_ID,
    DEFAULT_SURFACE_ID,
    GENERATE_A2UI_TOOL_NAME,
    GENERATE_A2UI_TOOL_DESCRIPTION,
    RENDER_A2UI_TOOL_DEF,
    build_a2ui_envelope,
    prepare_a2ui_request,
    wrap_error_envelope,
)


# Re-export the toolkit constants for callers that previously imported them
# from this package — keeps the public surface stable.
__all__ = [
    "get_a2ui_tools",
    "A2UI_OPERATIONS_KEY",
    "BASIC_CATALOG_ID",
]


def get_a2ui_tools(
    model: BaseChatModel,
    *,
    composition_guide: Optional[str] = None,
    default_surface_id: str = DEFAULT_SURFACE_ID,
    default_catalog_id: str = BASIC_CATALOG_ID,
    tool_name: str = GENERATE_A2UI_TOOL_NAME,
    tool_description: Optional[str] = None,
):
    """Build a LangGraph tool that delegates A2UI surface generation to a subagent.

    The returned tool is decorated with ``@langchain.tools.tool`` and is
    ready to bind into a chat model alongside any other tools.

    Args:
        model: Chat model the subagent will invoke for structured A2UI output.
            Using the same provider/model as the main agent is fine.
        composition_guide: Optional extra rules appended to the subagent's
            system prompt (e.g. project-specific component usage rules).
        default_surface_id: Surface id used when the subagent omits ``surfaceId``.
        default_catalog_id: Catalog id used when the subagent omits ``catalogId``.
        tool_name: Name advertised to the main agent's planner.
        tool_description: Description shown to the main agent's planner.

    Returns:
        A LangGraph tool callable suitable for ``bind_tools(...)``.
    """

    description = tool_description or GENERATE_A2UI_TOOL_DESCRIPTION

    @tool(tool_name, description=description)
    def generate_a2ui(
        runtime: ToolRuntime[Any],
        intent: str = "create",
        target_surface_id: Optional[str] = None,
        changes: Optional[str] = None,
    ) -> str:
        """Generate or edit an A2UI surface.

        Args:
            intent: Either ``"create"`` to render a new surface, or ``"update"``
                to modify a surface previously rendered in this conversation.
            target_surface_id: Required when ``intent="update"``. The surface
                id of the prior render to modify.
            changes: Optional natural-language description of the changes to
                apply when ``intent="update"``.
        """
        messages = runtime.state["messages"][:-1]

        # Shared: decide create/update, find prior surface, build the prompt.
        prep = prepare_a2ui_request(
            intent=intent,
            target_surface_id=target_surface_id,
            changes=changes,
            messages=messages,
            state=runtime.state,
            composition_guide=composition_guide,
        )
        if prep.get("error"):
            return wrap_error_envelope(prep["error"])

        # Glue: bind the structured-output tool and invoke the subagent.
        model_with_tool = model.bind_tools(
            [RENDER_A2UI_TOOL_DEF], tool_choice="render_a2ui"
        )
        response = model_with_tool.invoke(
            [SystemMessage(content=prep["prompt"]), *messages]
        )
        if not response.tool_calls:
            return wrap_error_envelope("LLM did not call render_a2ui")

        # Shared: assemble the final operations envelope.
        return build_a2ui_envelope(
            args=response.tool_calls[0]["args"],
            is_update=prep["is_update"],
            target_surface_id=target_surface_id,
            prior=prep["prior"],
            default_surface_id=default_surface_id,
            default_catalog_id=default_catalog_id,
        )

    return generate_a2ui

File: integrations/langgraph/typescript/src/a2ui-tool.ts
/**
 * A2UI subagent tool factory for LangGraph TS agents.
 *
 * Thin adapter over ``@ag-ui/a2ui-toolkit`` — the heavy lifting (op builders,
 * prompt assembly, history walkers, output envelope) lives in the toolkit so
 * each new framework adapter (ADK, Mastra, Strands, …) only owns the
 * framework-specific glue: tool decorator, runtime state access, model
 * binding + invoke.
 *
 * Example usage in a chat node:
 *
 *   import { getA2UITools } from "@ag-ui/langgraph";
 *
 *   const a2ui = getA2UITools(new ChatOpenAI({ model: "gpt-4o" }));
 *
 *   const modelWithTools = chatModel.bindTools(
 *     [...state.tools, a2ui],
 *     { parallel_tool_calls: false },
 *   );
 */

import { tool, type ToolRuntime } from "@langchain/core/tools";
import { SystemMessage } from "@langchain/core/messages";
import {
  A2UI_OPERATIONS_KEY,
  BASIC_CATALOG_ID,
  DEFAULT_SURFACE_ID,
  GENERATE_A2UI_TOOL_NAME,
  GENERATE_A2UI_TOOL_DESCRIPTION,
  GENERATE_A2UI_ARG_DESCRIPTIONS,
  RENDER_A2UI_TOOL_DEF,
  buildA2UIEnvelope,
  prepareA2UIRequest,
  wrapErrorEnvelope,
  runA2UIGenerationWithRecovery,
  type A2UIRecoveryConfig,
  type A2UIValidationCatalog,
  type A2UIAttemptRecord,
} from "@ag-ui/a2ui-toolkit";

/**
 * Loose type for the subagent model.
 *
 * Typed as `any` (rather than `BaseChatModel`) to tolerate `@langchain/core` version
 * skew between this package and the consumer — e.g. `ChatOpenAI` shipping its own
 * peer-pinned core. The factory only needs `bindTools` + `invoke`, which is checked
 * at runtime.
 */
export type A2UISubagentModel = any;

// Re-export the toolkit constants for callers that previously imported them
// from this package — keeps the public surface stable.
export { A2UI_OPERATIONS_KEY, BASIC_CATALOG_ID };

export interface A2UISubagentToolOptions {
  /** Optional extra rules appended to the subagent's system prompt. */
  compositionGuide?: string;
  /** Surface id used when the subagent omits `surfaceId`. */
  defaultSurfaceId?: string;
  /** Catalog id assigned to every new surface this factory creates — the
   *  subagent never picks the catalog. Falls back to the basic v0.9 catalog. */
  defaultCatalogId?: string;
  /** Name advertised to the main agent's planner. */
  toolName?: string;
  /** Description shown to the main agent's planner. */
  toolDescription?: string;
  /**
   * Inline catalog (component name → JSON Schema with `required`) enabling
   * catalog-aware recovery (unknown-component / missing-required-prop). Pass the
   * SAME catalog the host gives `@ag-ui/a2ui-middleware` so the retry decision
   * (here) and the paint gate (middleware) agree. Omit for structural-only
   * recovery (malformed JSON, missing root, broken refs/bindings).
   */
  catalog?: A2UIValidationCatalog;
  /** Recovery loop config: attempt cap, retry-UI threshold, debug exposure. */
  recovery?: A2UIRecoveryConfig;
  /** Per-attempt hook for emitting recovery status / dev logs (non-disruptive). */
  onA2UIAttempt?: (record: A2UIAttemptRecord) => void;
}

/** Tool arguments exposed to the main agent's planner. */
interface GenerateA2UIArgs {
  /**
   * `"create"` to render a new surface, `"update"` to modify a surface
   * previously rendered in this conversation. Defaults to `"create"`.
   */
  intent?: "create" | "update";
  /**
   * Required when `intent="update"`. The surface id of the prior render
   * to modify.
   */
  target_surface_id?: string;
  /** Optional natural-language description of the changes to apply on update. */
  changes?: string;
}

/**
 * Build a LangGraph tool that delegates A2UI surface generation to a subagent.
 *
 * The returned tool is ready to bind into a chat model alongside any other tools.
 *
 * @param model Chat model the subagent will invoke for structured A2UI output.
 *   Using the same provider/model as the main agent is fine.
 * @param options Optional behavior overrides.
 */
export function getA2UITools(
  model: A2UISubagentModel,
  options: A2UISubagentToolOptions = {},
) {
  // Use `||` rather than destructuring defaults so empty-string overrides fall
  // back to the canonical defaults (matches the Python adapter, which uses
  // `or` for the same parity). Otherwise an accidental `""` from a caller
  // would advertise a nameless / empty-description tool to the planner.
  const {
    compositionGuide,
    defaultSurfaceId: defaultSurfaceIdOpt,
    defaultCatalogId: defaultCatalogIdOpt,
    toolName: toolNameOpt,
    toolDescription: toolDescriptionOpt,
    catalog,
    recovery,
    onA2UIAttempt,
  } = options;
  const defaultSurfaceId = defaultSurfaceIdOpt || DEFAULT_SURFACE_ID;
  const defaultCatalogId = defaultCatalogIdOpt || BASIC_CATALOG_ID;
  const toolName = toolNameOpt || GENERATE_A2UI_TOOL_NAME;
  const toolDescription = toolDescriptionOpt || GENERATE_A2UI_TOOL_DESCRIPTION;

  return tool(
    async (
      input: GenerateA2UIArgs,
      runtime: ToolRuntime<Record<string, unknown>, unknown>,
    ): Promise<string> => {
      // Defensive: a custom state schema (or a non-graph invocation) may not
      // preseed `state`/`messages` — mirror the Python adapter's graceful
      // degrade (`state.get("messages", [])`) instead of throwing mid-tool.
      const state = (runtime.state ?? {}) as Record<string, unknown>;
      const allMessages = (state.messages as Array<any>) ?? [];
      // Strip current (unbalanced) tool call from history.
      const messages = allMessages.slice(0, -1);

      // Shared: decide create/update, find prior surface, build the prompt.
      const prep = prepareA2UIRequest({
        intent: input.intent,
        targetSurfaceId: input.target_surface_id,
        changes: input.changes,
        messages,
        state,
        compositionGuide,
      });
      if (prep.error) return wrapErrorEnvelope(prep.error);

      // Glue: bind the structured-output tool.
      if (!model.bindTools) {
        return wrapErrorEnvelope("Provided model does not support bindTools");
      }
      const modelWithTool = model.bindTools([RENDER_A2UI_TOOL_DEF], {
        tool_choice: { type: "function", function: { name: "render_a2ui" } },
      });

      // Shared: validate→retry loop. On each retry the prompt is re-augmented
      // with the prior attempt's structured errors; only a validated surface is
      // committed (the middleware gate suppresses any unvalidated attempt, so a
      // rejected attempt never paints). Returns a structured hard-failure
      // envelope once the attempt cap is hit.
      const { envelope } = await runA2UIGenerationWithRecovery({
        basePrompt: prep.prompt,
        catalog,
        config: recovery,
        onAttempt: onA2UIAttempt,
        invokeSubagent: async (prompt) => {
          const response: any = await modelWithTool.invoke([
            new SystemMessage(prompt),
            ...messages,
          ] as any);
          const toolCalls: Array<{ args?: Record<string, unknown> }> =
            response.tool_calls ?? [];
          return toolCalls.length ? (toolCalls[0].args ?? {}) : null;
        },
        buildEnvelope: (args) =>
          buildA2UIEnvelope({
            args,
            isUpdate: prep.isUpdate,
            targetSurfaceId: input.target_surface_id,
            prior: prep.prior,
            defaultSurfaceId,
            defaultCatalogId,
          }),
      });
      return envelope;
    },
    {
      name: toolName,
      description: toolDescription,
      schema: {
        type: "object",
        properties: {
          intent: {
            type: "string",
            enum: ["create", "update"],
            description: GENERATE_A2UI_ARG_DESCRIPTIONS.intent,
          },
          target_surface_id: {
            type: "string",
            description: GENERATE_A2UI_ARG_DESCRIPTIONS.target_surface_id,
          },
          changes: {
            type: "string",
            description: GENERATE_A2UI_ARG_DESCRIPTIONS.changes,
          },
        },
      } as any,
    },
  );
}
